pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.routing;
+
+import java.util.Map;
+import org.apache.pinot.common.request.BrokerRequest;
+import org.apache.pinot.core.transport.ServerInstance;
+
+/**
+ * The {@code RouteManager} provides the routing information for a query that requests access to a Pinot table.
+ *
+ * The implementation of this interface should ensure the routing and server information are up-to-date at the
+ * time when the routing request was made.
+ *
+ * set by the user. This needs to be added to support features like segment pruning.
+ */
+public interface RouteManager {
+
+  /**
+   * Validate routing exist for a table
+   *
+   * @param tableNameWithType the name of the table.
+   * @return true if the route table exists.
+   */
+  boolean routingExists(String tableNameWithType);
+
+  /**
+   * Get all enabled server instances that are available for routing.
+   *
+   * @return all currently enabled server instances.
+   */
+  Map<String, ServerInstance> getEnabledServerInstanceMap();
+
+  /**
+   * Get the {@link RouteTable} for a specified table.
+   *
+   * @param tableNameWithType the name of the table.
+   * @return the route table.
+   */
+  RouteTable getRoutingTable(String tableNameWithType);
```

Comment:
We can provide a default implementation to compile `SELECT * FROM <tableNameWithType>` and call `getRoutingTable()` with `brokerRequest`

Reply:
yeah I think i replied above. this basically 
- makes the code to create a SELECT * FROM <tableName> dummy query, generate a brokerRequest, and then gets the routing table. 
- then inside the getRouting table method it needs to parse the brokerrequest. parse out all the filter clauses (which is none) then apply pruning logic which is again unnecessary, then return back the full routing table

These are all unnecessary overheads to maintain correctness.

Reply:
I think this is because we are thinking of a simple JOIN query where the JOIN has to be performed by considering all segments of both tables so may be that's why we are thinking of a SELECT * type query in which case a dummy / default implementation of `getRoutingTable`() using `tableNameWithType` makes sene

However, consider a colocated join for two tables partitioned on the join key, the partition pruning has to happen / pushed below the join so the segment selector followed by instance selector are aware of the exact pruned segments that are going to participate in JOIN. 

In that case, we need the existing generic `BrokerRequest` based `getRoutingTable()` API imo.

Reply:
@walterddr  looks like this concern will be addressed based on the latest discussion that a `StageNode` based `getRoutingTable` API can be added in the future ?